sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsReadableSpan.java

```diff
@@ -190,6 +192,13 @@ public SpanData toSpanData() {
     return result;
   }
 
+  @Override
+  public boolean hasEnded() {
+    synchronized (lock) {
```

Comment:
this proliferation of synchronization really makes me want to implement this with AtomicBoolean, since it's lock-free.

Reply:
You could remove the lock here then, but all the checks for `hasEnded` within span still need to hold the lock if they want to ensure that they don't set properties after the span has ended.

Reply:
yeah. IMO, the whole thread-safety of this class is pretty janky. We should think about separating the mutable parts from the immutable and having a simpler synchronization model.